hub.go

```diff
@@ -208,75 +247,234 @@ func checkCmd(m []byte) {
 	} else if strings.HasPrefix(sl, "log off") {
 		*logDump = "off"
 		log.SetOutput(os.Stderr)
 		// } else if strings.HasPrefix(sl, "log show") {
 		// TODO: send all the saved log to websocket
-		//h.broadcastSys <- []byte("{\"BufFlowDebug\" : \"" + *logDump + "\"}")
+		//hub.broadcastSys <- []byte("{\"BufFlowDebug\" : \"" + *logDump + "\"}")
 	}
 }
 
-func memoryStats() {
+func (u *logWriter) Write(p []byte) (n int, err error) {
+	u.onWrite(p)
+	return len(p), nil
+}
+
+func (hub *hub) memoryStats() {
 	var memStats runtime.MemStats
 	runtime.ReadMemStats(&memStats)
 	json, _ := json.Marshal(memStats)
 	log.Printf("memStats:%v\n", string(json))
-	h.broadcastSys <- json
+	hub.broadcastSys <- json
 }
 
-func getHostname() {
-	h.broadcastSys <- []byte("{\"Hostname\" : \"" + *hostname + "\"}")
+func (hub *hub) getHostname() {
+	hub.broadcastSys <- []byte("{\"Hostname\" : \"" + *hostname + "\"}")
 }
 
-func getVersion() {
-	h.broadcastSys <- []byte("{\"Version\" : \"" + version + "\"}")
+func (hub *hub) getVersion() {
+	hub.broadcastSys <- []byte("{\"Version\" : \"" + version + "\"}")
 }
 
-func garbageCollection() {
+func (hub *hub) garbageCollection() {
 	log.Printf("Starting garbageCollection()\n")
-	h.broadcastSys <- []byte("{\"gc\":\"starting\"}")
-	memoryStats()
+	hub.broadcastSys <- []byte("{\"gc\":\"starting\"}")
+	hub.memoryStats()
 	debug.SetGCPercent(100)
 	debug.FreeOSMemory()
 	debug.SetGCPercent(-1)
 	log.Printf("Done with garbageCollection()\n")
-	h.broadcastSys <- []byte("{\"gc\":\"done\"}")
-	memoryStats()
+	hub.broadcastSys <- []byte("{\"gc\":\"done\"}")
+	hub.memoryStats()
+}
+
+func (hub *hub) spHandlerOpen(portname string, baud int, buftype string) {
+
+	log.Print("Inside spHandler")
+
+	var out bytes.Buffer
+
+	out.WriteString("Opening serial port ")
+	out.WriteString(portname)
+	out.WriteString(" at ")
+	out.WriteString(strconv.Itoa(baud))
+	out.WriteString(" baud")
+	log.Print(out.String())
+
+	conf := &SerialConfig{Name: portname, Baud: baud, RtsOn: true}
+
+	mode := &serial.Mode{
+		BaudRate: baud,
+	}
+
+	sp, err := serial.Open(portname, mode)
+	log.Print("Just tried to open port")
+	if err != nil {
+		//log.Fatal(err)
+		log.Print("Error opening port " + err.Error())
+		//hub.broadcastSys <- []byte("Error opening port. " + err.Error())
+		hub.broadcastSys <- []byte("{\"Cmd\":\"OpenFail\",\"Desc\":\"Error opening port. " + err.Error() + "\",\"Port\":\"" + conf.Name + "\",\"Baud\":" + strconv.Itoa(conf.Baud) + "}")
+
+		return
+	}
+	log.Print("Opened port successfully")
+	//p := &serport{send: make(chan []byte, 256), portConf: conf, portIo: sp}
+	// we can go up to 256,000 lines of gcode in the buffer
+	p := &serport{
+		sendBuffered: make(chan string, 256000),
+		sendNoBuf:    make(chan []byte),
+		sendRaw:      make(chan string),
+		portConf:     conf,
+		portIo:       sp,
+		portName:     portname,
+		BufferType:   buftype,
+	}
+
+	p.OnMessage = func(msg []byte) {
+		hub.broadcastSys <- msg
+	}
```

Comment:
Instead of this call back I think you could use the `ChanWriter` I suggested above
```suggestion
	p := &serport{
		sendBuffered: make(chan string, 256000),
		sendNoBuf:    make(chan []byte),
		sendRaw:      make(chan string),
		portConf:     conf,
		portIo:       sp,
		portName:     portname,
		BufferType:   buftype,
		Writer: &ChanWriter{Ch: hub.broadcastSys},
	}
```